cloud/util/util.go

```diff
@@ -254,3 +261,136 @@ func CreateClientProviderFromClusterIdentity(ctx context.Context, client client.
+				helper, err := patch.NewHelper(&actualMachine, params.Client)
+				if err != nil {
+					return err
+				}
+				actualMachine.Status.Ready = true
+				err = helper.Patch(ctx, &actualMachine)
+				if err != nil {
+					return err
+				}
+			}
+			continue
+		}
+
+		labels := map[string]string{
+			clusterv1.ClusterNameLabel:     params.Cluster.Name,
+			clusterv1.MachinePoolNameLabel: format.MustFormatValue(params.MachinePool.Name),
+		}
+		infraMachine := &infrav2exp.OCIMachinePoolMachine{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:    params.Namespace,
+				GenerateName: fmt.Sprintf("%s-", params.InfraMachinePoolName),
+				Labels:       labels,
+				Annotations:  make(map[string]string),
+				// set the parent to infra machinepool till the capi machine reconciler changes it to capi machinepool machine
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						Kind:       params.InfraMachinePoolKind,
+						Name:       params.InfraMachinePoolName,
+						APIVersion: infrav2exp.GroupVersion.String(),
+						UID:        params.InfraMachinePoolUID,
+					},
+				},
+			},
+			Spec: infrav2exp.OCIMachinePoolMachineSpec{
+				OCID:         specMachine.Spec.OCID,
+				ProviderID:   specMachine.Spec.ProviderID,
+				InstanceName: specMachine.Spec.InstanceName,
+				MachineType:  specMachine.Spec.MachineType,
+			},
+		}
+		infraMachine.Status.Ready = specMachine.Status.Ready
+		controllerutil.AddFinalizer(infraMachine, infrav2exp.MachinePoolMachineFinalizer)
+		params.Logger.Info("Creating machinepool  machine", "machine", infraMachine.Name, "instanceName", specMachine.Name)
+
+		if err := params.Client.Create(ctx, infraMachine); err != nil {
+			return errors.Wrap(err, "failed to create machine")
+		}
+	}
+
+	return nil
+}
+
+func getMachinepoolMachines(ctx context.Context, c client.Client, machinePool *expclusterv1.MachinePool, cluster *clusterv1.Cluster, namespace string) (*infrav2exp.OCIMachinePoolMachineList, error) {
+	machineList := &infrav2exp.OCIMachinePoolMachineList{}
+	labels := map[string]string{
+		clusterv1.ClusterNameLabel:     cluster.Name,
+		clusterv1.MachinePoolNameLabel: format.MustFormatValue(machinePool.Name),
+	}
+	if err := c.List(ctx, machineList, client.InNamespace(namespace), client.MatchingLabels(labels)); err != nil {
+		return nil, err
+	}
+
+	return machineList, nil
+}
+
+// DeleteOrphanedMachinePoolMachines deletes the machine pool machines which are not required. This method lists the
+// existing machines in the clusters and does a diff with the spec and deletes any machines which are missing from the spec.
+func DeleteOrphanedMachinePoolMachines(ctx context.Context, params MachineParams) error {
+	machineList, err := getMachinepoolMachines(ctx, params.Client, params.MachinePool, params.Cluster, params.Namespace)
+	if err != nil {
+		return err
+	}
+
+	instanceNameSet := map[string]struct{}{}
+	for _, specMachine := range params.SpecInfraMachines {
+		instanceNameSet[*specMachine.Spec.OCID] = struct{}{}
```

Comment:
Is all this does this way simply as a "does this id exist" tests on line 364?

Reply:
yeah, in order to delete a machine, we need to check if it isn the spec, so we create a map and see if the key is in the map, since go does not have set. Some of this is copied from docker machine pool machine PR , so if you have any suggestions, please feel free to comment :).